dace/transformation/helpers.py

```diff
@@ -17,6 +18,307 @@
+                out_state.add_edge(tasklet, '__out', acc, None, Memlet.from_array(nname, ndesc))
+                write_set.add(name)
+
+        # Add NestedSDFG node
+        fsymbols = sdfg.symbols.keys() | nsdfg.free_symbols
+        fsymbols.update(defined_symbols - strictly_defined_symbols)
+        mapping = {s: s for s in fsymbols}
+        cnode = new_state.add_nested_sdfg(nsdfg, None, read_set, write_set, mapping)
+        for s in strictly_defined_symbols:
+            if s in sdfg.symbols:
+                sdfg.remove_symbol(s)
+
+        for name in read_set:
+            r = new_state.add_read(name)
+            new_state.add_edge(r, None, cnode, name, Memlet.from_array(name, sdfg.arrays[name]))
+        for name in write_set:
+            w = new_state.add_write(name)
+            new_state.add_edge(cnode, name, w, None, Memlet.from_array(name, sdfg.arrays[name]))
+
+        if out_state is not None:
+            extra_state = sdfg.add_state('symbolic_output')
+            for e in sdfg.out_edges(new_state):
+                sdfg.add_edge(extra_state, e.dst, e.data)
+                sdfg.remove_edge(e)
+            sdfg.add_edge(new_state, extra_state, InterstateEdge(assignments=out_mapping))
+            new_state = extra_state
+
+    else:
+        new_state = states[0]
+
+    return new_state
+
+
+def _copy_state(sdfg: SDFG,
+                state: SDFGState,
+                before: bool = True,
+                states: Optional[Set[SDFGState]] = None) -> SDFGState:
+
+    state_copy = copy.deepcopy(state)
+    state_copy._label += '_copy'
+    sdfg.add_node(state_copy)
+
+    in_conditions = []
+    for e in sdfg.in_edges(state):
+        if states and e.src not in states:
+            continue
+        sdfg.add_edge(e.src, state_copy, e.data)
+        sdfg.remove_edge(e)
+        if not e.data.is_unconditional():
+            in_conditions.append(e.data.condition.as_string)
+
+    out_conditions = []
+    for e in sdfg.out_edges(state):
+        if states and e.dst not in states:
+            continue
+        sdfg.add_edge(state_copy, e.dst, e.data)
+        sdfg.remove_edge(e)
+        if not e.data.is_unconditional():
+            out_conditions.append(e.data.condition.as_string)
+
+    if before:
+        condition = None
+        if in_conditions:
+            condition = 'or'.join([f"({c})" for c in in_conditions])
+        sdfg.add_edge(state_copy, state, InterstateEdge(condition=condition))
+    else:
+        condition = None
+        # NOTE: The following should be unecessary for preserving program semantics. Therefore we comment it out to
+        # avoid the overhead of evaluating the condition.
+        # if out_conditions:
+        #     condition = 'or'.join([f"({c})" for c in out_conditions])
+        sdfg.add_edge(state, state_copy, InterstateEdge(condition=condition))
+
+    return state_copy
+
+
+def find_sdfg_control_flow(sdfg: SDFG):
+
+    split_interstate_edges(sdfg)
```

Comment:
deepcopy? would be slow so no need but I'm wondering why we do this

Reply:
Experimental. The expected use of the method is to copy scope source/sink states (this is why it is private in the first place) that belong to multiple scopes (direct result of cf.structured_control_flow_tree). I would expect such states to be always empty, but deep-copying to be safe. Can be amended in the future.